babel.config.js

```diff
@@ -25,52 +25,218 @@
+  localPlugin('transform-version-call'),
+  localPlugin('transform-simple-array-destructure'),
+  getReplacePlugin(isEsmBuild),
 ];
 
-const esmPresets = [
-  'babel-preset-modules',
-  {
-    'loose': true,
-  },
-];
-const presets = [
-  '@babel/preset-env',
-  {
-    'modules': isClosureCompiler ? false : 'commonjs',
-    'loose': true,
-    'targets': noModuleTarget,
-  },
+const esmRemovedImports = {
+  './polyfills/document-contains': ['installDocContains'],
+  './polyfills/domtokenlist': ['installDOMTokenList'],
+  './polyfills/fetch': ['installFetch'],
+  './polyfills/math-sign': ['installMathSign'],
+  './polyfills/object-assign': ['installObjectAssign'],
+  './polyfills/object-values': ['installObjectValues'],
+  './polyfills/promise': ['installPromise'],
+  './polyfills/array-includes': ['installArrayIncludes'],
+};
+
+// Removable imports that are not needed for valid transformed documents.
+const validTransformedRemovableImports = {
+  './build-system/build/ampshared.css': ['cssText', 'ampSharedCss'],
+  './build-system/build/ampdoc.css': ['cssText', 'ampDocCss'],
+};
+
+/**
+ * @param {boolean} isEsmBuild a boolean indicating if this build is for ESM output.
+ * @return {Array<string|Object>} the minify-replace plugin options that can be
+ * pushed into the babel plugins array
+ */
+function getReplacePlugin(isEsmBuild) {
+  /**
+   * @param {string} identifierName the identifier name to replace
+   * @param {boolean} value the value to replace with
+   * @return {!Object} replacement options used by minify-replace plugin
+   */
+  function createReplacement(identifierName, value) {
+    return {
+      identifierName,
+      replacement: {
+        type: 'booleanLiteral',
+        value,
+      },
+    };
+  }
+
+  const replacements = [createReplacement('IS_ESM', isEsmBuild)];
+  const defineFlag = argv.defineExperimentConstant;
+
+  // add define flags from arguments
+  if (Array.isArray(defineFlag)) {
+    if (defineFlag.length > 1) {
+      throw new Error('Only one defineExperimentConstant flag is allowed');
+    } else {
+      replacements.push(createReplacement(defineFlag[0], true));
+    }
+  } else if (defineFlag) {
+    replacements.push(createReplacement(defineFlag, true));
+  }
+
+  // default each experiment flag constant to false
+  Object.keys(experimentsConfig).forEach(experiment => {
+    const experimentDefine =
+      experimentsConfig[experiment]['defineExperimentConstant'];
+
+    // eslint-disable-next-line require-jsdoc
+    function flagExists(element) {
+      return element['identifierName'] === experimentDefine;
+    }
+
+    // only add default replacement if it already doesn't exist in array
+    if (experimentDefine && !replacements.some(flagExists)) {
+      replacements.push(createReplacement(experimentDefine, false));
+    }
+  });
+
+  // default each backup experiment constant to the customized value
+  // eslint-disable-next-line local/no-for-of-statement, local/no-array-destructuring
```

Comment:
The directory `build-system/compile` in which `build.conf.js` lived had a much more permissive `.eslintrc`. I didn't try to rewrite things here to fit the stricter rule set that applies to this file, and for now have just added `eslint-disable-next-line` where relevant.

Reply:
Instead of declaring exceptions via inline comments, you can add overrides specific to `babel.config.js` here:

https://github.com/ampproject/amphtml/blob/3488672872a98df3df29a150bffc7bfca7994586/.eslintrc#L259-L266

In addition, if @jridgewell thinks there's good enough reason to rewrite code to remove those rule exceptions, we can do so via a separate follow up. (Maybe add a TODO somewhere in this PR?)